app/models/comment.rb

```diff
@@ -32,6 +32,7 @@ class Comment < ApplicationRecord
   has_many :mentions, as: :mentionable, inverse_of: :mentionable, dependent: :destroy
   has_many :notifications, as: :notifiable, inverse_of: :notifiable, dependent: :delete_all
   has_many :notification_subscriptions, as: :notifiable, inverse_of: :notifiable, dependent: :destroy
+  has_many :edits, class_name: "CommentEdit", dependent: :nullify
```

Comment:
having nullified the comment_id on the comment_edits - how would you know what the comment edit was for? Are we losing context when a comment is deleted? 

Comment's are soft-deleted if they have children, but destroyed otherwise, so this would not be an uncommon event.

Since we're not adding a comment edit for creation (where id would be a new fact) and deletion - there's a common case of non-edited comments (where the initial version is abusive, and then it's deleted after some pushback from the community or self-policing) and we have no comment edit

I _think_ we want to keep the created version as a comment edit as well - and I'm not sure we need or want to nullify the comment_id (since that provides a useful key to thread these together, in the strange case where a user was editing in multiple comment threads at once)

I tested this locally and I get a history like this

```
# comment created - no edits
[4] pry(main)> CommentEdit.count
   (1.1ms)  SELECT COUNT(*) FROM "comment_edits"
=> 0

# comment edited once - comment id present on the edit:
[5] pry(main)> CommentEdit.count
   (0.8ms)  SELECT COUNT(*) FROM "comment_edits"
=> 1
[6] pry(main)> CommentEdit.all
  CommentEdit Load (1.7ms)  SELECT "comment_edits".* FROM "comment_edits"
=> [#<CommentEdit:0x000055bca5de4768
  id: 1,
  comment_id: 45,
  modifications:
   {"edited_at"=>{"to"=>"2021-08-30T19:12:06.246Z", "from"=>nil},
    "body_markdown"=>
     {"to"=>"comment 1 is a comment\r\n\r\nFilthy abusive language goes here.",
      "from"=>"comment 1 is a comment"}},
  editor_id: 11,
  created_at: Mon, 30 Aug 2021 19:12:06.370610000 UTC +00:00,
  updated_at: Mon, 30 Aug 2021 19:12:06.370610000 UTC +00:00>]

# comment deleted - there were no child comments so this leaf is destroyed rather than marked deleted - comment_id removed from the comment_edit:
[7] pry(main)> CommentEdit.all
  CommentEdit Load (1.8ms)  SELECT "comment_edits".* FROM "comment_edits"
=> [#<CommentEdit:0x000055bca5d0f040
  id: 1,
  comment_id: nil,
  modifications:
   {"edited_at"=>{"to"=>"2021-08-30T19:12:06.246Z", "from"=>nil},
    "body_markdown"=>
     {"to"=>"comment 1 is a comment\r\n\r\nFilthy abusive language goes here.",
      "from"=>"comment 1 is a comment"}},
  editor_id: 11,
  created_at: Mon, 30 Aug 2021 19:12:06.370610000 UTC +00:00,
  updated_at: Mon, 30 Aug 2021 19:12:06.370610000 UTC +00:00>]
```

Non-blocking but it's possible this isn't fulfilling the goal.

Reply:
> having nullified the comment_id on the comment_edits - how would you know what the comment edit was for? Are we losing context when a comment is deleted? … and I'm not sure we need or want to nullify the comment_id (since that provides a useful key to thread these together, in the strange case where a user was editing in multiple comment threads at once)

This wasn't much of a conscious decision, tbh. The linter required a `dependent` option, so I added one just so I could commit it. 🙂 I didn't know if we wanted to delete the edits if a comment was deleted, and since comments aren't currently deleted when articles are, I didn't know what made the most sense here. I also didn't look up what all options there are for `dependent` here, so I chose the only one I knew of off the top of my head that didn't delete the record in the DB.

I forgot I'd done this, though, so it's probably a good thing you called it out.

> versions were indexed on object id and object type

Speaking of which, I should probably add the FK indexes. I meant to do that but completely forgot. I had to pause to do something else and lost context.

Reply:
could we use `:dependent nil` instead (do nothing) - not sure if that makes the cop happy but it prevents removing the foreign key

I know I'm advocating for data inconsistency here, but I really think user initiated comment deletion should not remove the link to the history.

I'm also 100% on board with revisiting the [logic in the controller](https://github.com/forem/forem/blob/main/app/controllers/comments_controller.rb#L186-L191) to always soft delete comments (as we do for comments with children or admin deletions)  - then the comment destroy only happens on user deletion (via `DeleteComments.call(user)`) at which point we're not concerned with moderation - and any user deleted comment still has its edits and its primary key (but is no longer shown in the article view).